scst/src/scst_dlm.c

```diff
@@ -1494,6 +1615,199 @@ static void scst_pr_dlm_cleanup(struct scst_device *dev)
+	scst_dlm_free_rem_ua(ua);
+
+	TRACE_EXIT();
+}
+
+/*
+ * scst_dlm_pr_reg_send_rem_ua - send a unit attention to the registrant on a remote node
+ *
+ * This will create a lock containing the necessary UA information, increment next_rem_ua_idx and
+ * add the struct scst_dlm_rem_ua to the registrant's sent_rem_ua_list.
+ */
+static void scst_dlm_pr_reg_send_rem_ua(struct scst_device *dev, dlm_lockspace_t *ls,
+					struct scst_dev_registrant *reg, struct scst_dlm_rem_ua *ua)
+{
+	char reg_ua_name[32];
+	struct dlm_ua_lvb *ua_lvb;
+
+	TRACE_ENTRY();
+
+	snprintf(reg_ua_name, sizeof(reg_ua_name), PR_REG_UA_LOCK, reg->dlm_idx,
+		 reg->next_rem_ua_idx);
+
+	/* Always create a new lock. */
+	if (scst_dlm_lock_wait(ls, DLM_LOCK_NL, &ua->lksb, 0, reg_ua_name, NULL) < 0) {
+		PRINT_ERROR("Failed to lock (NL) %s.%s", dev->virt_name, reg_ua_name);
+		return;
+	}
+
+	if (scst_dlm_lock_wait(ls, DLM_LOCK_PW, &ua->lksb,
+			       DLM_LKF_VALBLK | DLM_LKF_CONVERT,
+			       reg_ua_name, NULL) >= 0) {
+		ua_lvb = (void *)ua->lksb.lksb.sb_lvbptr;
+
+		memset(ua->lvb, 0, sizeof(ua->lvb));
+		ua_lvb->version = 1;
+		ua_lvb->key = ua->key;
+		ua_lvb->asc = ua->asc;
+		ua_lvb->ascq = ua->ascq;
+
+		scst_dlm_lock_wait(ls, DLM_LOCK_CR, &ua->lksb,
+				   DLM_LKF_CONVERT | DLM_LKF_VALBLK,
+				   reg_ua_name, NULL);
+		reg->next_rem_ua_idx++;
+		list_add_tail(&ua->rem_ua_list_entry, &reg->sent_rem_ua_list);
+	} else {
+		PRINT_ERROR("Failed to lock %s.%s", dev->virt_name, reg_ua_name);
+	}
+
+	TRACE_EXIT();
+}
+
+/*
+ * scst_dlm_rm_rem_ua_ls - Unlock the lock associated with a remote unit attention
+ */
+static void scst_dlm_rm_rem_ua_ls(dlm_lockspace_t *ls, struct scst_dlm_rem_ua *ua)
+{
+	int res;
+
+	if (!ua->lksb.lksb.sb_lkid)
+		return;
+	res = scst_dlm_unlock_wait(ls, &ua->lksb);
+	WARN(res < 0, "scst_dlm_unlock_wait(%08x) failed (%d)",
+	     ua->lksb.lksb.sb_lkid, res);
+	ua->lksb.lksb.sb_lkid = 0;
+}
+
+/*
+ * scst_dlm_pr_reg_recv_rem_ua - receive a remotely generated UNIT ATTENTION for this registrant
+ *
+ * This will read unit attention that was 'published' by another node as a lock.  The resultant
+ * struct scst_dlm_rem_ua will NOT be inserted into a list on this node.
+ */
+static struct scst_dlm_rem_ua *scst_dlm_pr_reg_recv_rem_ua(struct scst_device *dev,
+							   dlm_lockspace_t *ls,
+							   struct scst_dev_registrant *reg,
+							   int index)
+{
+	char reg_ua_name[32];
+	struct scst_dlm_rem_ua *ua;
+	struct dlm_ua_lvb *ua_lvb;
+	int res;
+
+	ua = scst_dlm_alloc_rem_ua();
+	if (!ua)
+		return NULL;
+
+	ua_lvb = (void *)ua->lksb.lksb.sb_lvbptr;
+	snprintf(reg_ua_name, sizeof(reg_ua_name), PR_REG_UA_LOCK, reg->dlm_idx, index);
+
+	res = scst_dlm_lock_wait(ls, DLM_LOCK_PW, &ua->lksb,
+				 DLM_LKF_VALBLK, reg_ua_name, NULL);
+	if (res < 0) {
+		res = -EFAULT;
+		PRINT_ERROR("locking %s.%s failed", dev->virt_name,
+			    reg_ua_name);
+		goto cancel;
+	} else if (ua->lksb.lksb.sb_flags & DLM_SBF_VALNOTVALID) {
+		res = -EINVAL;
```

Comment:
In case when `scst_dlm_lock_wait()` returned success, should it be canceled here with  `scst_dlm_unlock_wait()`?

Reply:
Yes.  Thanks.